arcgis-ios-sdk-samples/Layers/List KML contents/ListKMLContentsViewController.swift

```diff
@@ -0,0 +1,113 @@
+        }
+        
+        // Add the source code button item to the right of navigation bar.
+        (navigationItem.rightBarButtonItem as! SourceCodeBarButtonItem).filenames = ["ListKMLContentsViewController", "ListKMLContentsSceneViewController"]
+        
+        // create the dataset from the local kml/kmz file with the given name
+        let kmlDataset = AGSKMLDataset(name: "esri_test_data")
+        self.kmlDataset = kmlDataset
+        
+        // load the dataset asynchronously so we can list its contents
+        kmlDataset.load {[weak self] (error) in
+            
+            guard let self = self else{
+                return
+            }
+            
+            guard error == nil else{
+                // display the error as an alert
+                let alertController = UIAlertController(title: error?.localizedDescription, message: nil, preferredStyle: .alert)
+                alertController.addAction(UIAlertAction(title: "OK", style: .default))
+                self.present(alertController, animated: true, completion: nil)
+                return
+            }
+            
+            self.flattenedNodes = []
+            func addNodes(_ nodes: [AGSKMLNode], level: Int){
+                for node in nodes{
+                    node.isVisible = true
+                    self.flattenedNodes.append((node, level))
+                    addNodes(self.childNodes(of: node), level: level+1)
+                }
+            }
+            addNodes(kmlDataset.rootNodes, level: 0)
+            
+            // populate the outline view now that the data is loaded
+            self.tableView.reloadData()
+        }
+    }
+    
+    /// Returns all the child nodes of this node.
+    private func childNodes(of node: AGSKMLNode) -> [AGSKMLNode] {
+        if let container = node as? AGSKMLContainer {
+            return container.childNodes
+        }
+        else if let networkLink = node as? AGSKMLNetworkLink {
+            return networkLink.childNodes
+        }
+        return []
+    }
+    
+    //MARK: - UITableViewDataSource
+    
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return flattenedNodes.count
+    }
+    
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        // get and setup a cell to display the node info
+        let cell = tableView.dequeueReusableCell(withIdentifier: "NodeCell")!
+        let flattenedNodeInfo = flattenedNodes[indexPath.row]
+        let node = flattenedNodeInfo.node
+        cell.textLabel?.text = node.name
+        cell.detailTextLabel?.text = "\(type(of:node))"
+        // set the indentation based on the node hierarchy
+        cell.indentationLevel = flattenedNodeInfo.level
+        return cell
+    }
+    
+    //MARK: - UITableViewDelegate
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
```

Comment:
If you wanted to, you could trigger the presentation via a segue configured in the storyboard. Then in `prepareForSegue()`, you would get the node corresponding to the table view's selected index path and assign that to the destination view controller. I think that method is more common, but what you have is just fine too.

Reply:
It's already working this way so I don't feel compelled to change it.